pertpy/tools/_sccoda.py

```diff
@@ -0,0 +1,2054 @@
+            kernel (npy.infer.mcmc.MCMCKernel): A `numpyro.infer.mcmc.MCMCKernel` object
+            rng_key (Array or random.PRNGKey): The rng state used. If None, a random state will be selected
+            copy (bool, optional): Return a copy instead of writing to adata. Defaults to False.
+            args: Passed to `numpyro.infer.mcmc.MCMC`
+            kwargs: Passed to `numpyro.infer.mcmc.MCMC`
+
+        Returns:
+            Saves all results into `sample_adata` and generates `self.mcmc` as a class attribute. If copy==True, return a copy of adata.
+        """
+        dtype = "float64"
+        # The tracked MCMC parameters for diagnostic checks
+        extra_fields = (
+            "potential_energy",
+            "num_steps",
+            "adapt_state.step_size",
+            "accept_prob",
+            "mean_accept_prob",
+        )
+
+        # Convert all data needed for modeling to numpyro arrays
+        numpyro_counts = jnp.array(sample_adata.X, dtype=dtype)
+        numpyro_covariates = jnp.array(sample_adata.obsm["covariate_matrix"], dtype=dtype)
+        numpyro_n_total = jnp.array(sample_adata.obsm["sample_counts"], dtype=dtype)
+
+        # Create mcmc attribute and run inference
+        self.mcmc = MCMC(kernel, *args, **kwargs)
+        self.mcmc.run(
+            rng_key,
+            numpyro_counts,
+            numpyro_covariates,
+            numpyro_n_total,
+            jnp.array(sample_adata.uns["scCODA_params"]["reference_index"]),
+            sample_adata,
+            init_params=sample_adata.uns["scCODA_params"]["mcmc"]["init_params"],
+            extra_fields=extra_fields,
+        )
+
+        # Set acceptance rate and save sampled values to `sample_adata.uns`
+        sample_adata.uns["scCODA_params"]["mcmc"]["acceptance_rate"] = self.mcmc.last_state.mean_accept_prob.to_py()
+        sample_adata.uns["scCODA_params"]["mcmc"]["samples"] = self.mcmc.get_samples()
+
+        # Evaluate results and create result dataframes (based on tre-aggregation or not)
+        if sample_adata.uns["scCODA_params"]["model_type"] == "classic":
+            intercept_df, effect_df = self.summary_prepare(sample_adata)  # type: ignore
+        elif sample_adata.uns["scCODA_params"]["model_type"] == "tree_agg":
+            intercept_df, effect_df, node_df = self.summary_prepare(sample_adata)  # type: ignore
+            # Save node df in `sample_adata.uns`
+            sample_adata.uns["scCODA_params"]["node_df"] = node_df
+        else:
+            raise ValueError("No valid model type!")
+
+        # Save intercept and effect dfs in `sample_adata.varm` (one effect df per covariate)
+        sample_adata.varm["intercept_df"] = intercept_df
+        for cov in effect_df.index.get_level_values("Covariate"):
+            sample_adata.varm[f"effect_df_{cov}"] = effect_df.loc[cov, :]
+        if copy:
+            return sample_adata
+
+    def go_nuts(
```

Comment:
Uhm, is this how the function should be named?

Reply:
Yeah. Johannes named it.

Reply:
Renamed to run_nuts